### PR TITLE
Rename  `bounded_order_preserving_variants` config to `prefer_exising_sort` and update docs

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

closes https://github.com/apache/arrow-datafusion/issues/7722

## Rationale for this change

While debugging an issue upgrading our code to use DataFuson, @ozankabak  pointed me at the following config: https://github.com/apache/arrow-datafusion/pull/7671#discussion_r1341636365

This setting (I think) controls if the DataFusion planner should prefer using the existing sort order or trying to maximize paralleilsm using repartition and re-sorting

It turns out to be the right one, but I don't think I would have found it without @ozankabak 's suggestion

I think the core of my challenge is that the current name describes how it modifies DataFusion's algorithms rather than what effect it has on the plans

## What changes are included in this PR?
I propose to change the config to `prefer_existing_sort` and update the documentation

## Are these changes tested?
existing tests

## Are there any user-facing changes?
yes, a config setting has a different name (and this is a breaking API change)

### CODE REVIEW

**Review by ozankabak — APPROVED** [review]
I like the new name, thanks @alamb

File: docs/source/user-guide/configs.md [hunk 1]
```diff
@@ -87,7 +87,7 @@ Environment variables are read during `SessionConfig` initialisation so they mus
...
+| datafusion.optimizer.prefer_existing_sort                  | false                     | When true, DataFusion will opportunistically remove sorts when the data is already sorted, replacing `RepartitionExec` with `SortPreservingRepartitionExec`, and `CoalescePartitionsExec` with `SortPreservingMergeExec`, When false, DataFusion will prefer to maximize the parallelism using `Repartition/Coalesce` and resort the data subsequently with `SortExec`                                                                                                                                                                                                                                  |
```
Comment (viirya) [inline docs/source/user-guide/configs.md:90]:
```suggestion
| datafusion.optimizer.prefer_existing_sort                  | false                     | When true, DataFusion will opportunistically remove sorts when the data is already sorted, replacing `RepartitionExec` with `SortPreservingRepartitionExec`, and `CoalescePartitionsExec` with `SortPreservingMergeExec`. When false, DataFusion will prefer to maximize the parallelism using `Repartition/Coalesce` and resort the data subsequently with `SortExec`                                                                                                                                                                                                                                  |
```

File: datafusion/common/src/config.rs [hunk 2]
```diff
@@ -453,11 +453,13 @@ config_namespace! {
         /// ```
         pub repartition_sorts: bool, default = true
 
-        /// When true, DataFusion will opportunistically remove sorts by replacing
-        /// `RepartitionExec` with `SortPreservingRepartitionExec`, and
+        /// When true, DataFusion will opportunistically remove sorts when the data is already sorted,
+        /// replacing `RepartitionExec` with `SortPreservingRepartitionExec`, and
```
Comment (viirya) [inline datafusion/common/src/config.rs:457]: Hmm, actually there is no `SortPreservingRepartitionExec` operator but it is a variant of `RepartitionExec` with `preserve_order` as true. It is a little confusion at first if trying to look for `SortPreservingRepartitionExec` type in IDE.

Maybe:

```suggestion
        /// replacing `RepartitionExec` with `SortPreservingRepartitionExec` (i.e., `RepartitionExec` with `preserve_order` as true), and
```

**Review by viirya — APPROVED** [review]

File: datafusion/common/src/config.rs (hunk 2 above)
Comment (alamb) [reply datafusion/common/src/config.rs]: That makes sense, thank you @viirya  --  I tried to improve the wording in 35c674852b. Let me know what you think

Comment (viirya) [reply datafusion/common/src/config.rs]: Looks good to me. Thanks @alamb

**Pull request merged**